### PR TITLE
Automatically include Content-Length

### PR DESCRIPTION
**Pull request by stephencelis**
Alternative (or supporting?) idea to https://github.com/pointfreeco/swift-web/pull/52.

### CODE REVIEW

File: Sources/HttpPipeline/Middleware.swift [hunk 1]
```diff
@@ -109,24 +110,25 @@ public func send(_ data: Data?) -> Middleware<BodyOpen, BodyOpen, Data?, Data?>
...
 
 public func respond<A>(json: String) -> Middleware<HeadersOpen, ResponseEnded, A, Data?> {
-  return map(const(json.data(using: .utf8)))
-    >>> writeHeader(.contentType(.json))
-    >>> closeHeaders
-    >>> end
+  return respond(body: json, contentType: .json)
+}
+
+private func respond<A>(body: String, contentType: MediaType)
```
Comment (stephencelis) [inline Sources/HttpPipeline/Middleware.swift]: This should maybe be public?

File: Tests/HttpPipelineTests/__Snapshots__/HttpPipelineTests/testBasicAuth_Authorized.1.Conn.txt [hunk 2]
```diff
@@ -13,10 +13,8 @@ Body: (Data, 0 bytes)
 
 Response:
 ---------
-Status 200 (ok)
-Headers: [
-  Content-Type: text/html; charset=utf8
-]
-Bytes: 19
+Status 200 OK
```
Comment (stephencelis) [inline Tests/HttpPipelineTests/__Snapshots__/HttpPipelineTests/testBasicAuth_Authorized.1.Conn.txt]: Having a little fun with the response formatting to make it look more like a raw HTTP response.

**Comment by mbrandonw** [comment]
> Having a little fun with the response formatting to make it look more like a raw HTTP response.

oh yeah I have an issue here to format it like HTTP/1.1 message format:

https://github.com/pointfreeco/swift-web/issues/43

File: Sources/HttpPipeline/Middleware.swift (hunk 1 above)
Comment (mbrandonw) [reply Sources/HttpPipeline/Middleware.swift]: yeah let's make it public!

File: Tests/HttpPipelineTests/HttpPipelineTests.swift [hunk 3]
```diff
@@ -84,7 +84,11 @@ class HttpPipelineTests: XCTestCase {
     let middleware: Middleware<StatusLineOpen, ResponseEnded, Prelude.Unit, Data?> =
       contentLength
         <| writeStatus(.ok)
-        >>> respond(html: "<p>Hello, world</p>")
+        >>> writeHeader(.contentType(.html))
+        >>> closeHeaders
+        >>> map(const(Data()))
```
Comment (stephencelis) [inline Tests/HttpPipelineTests/HttpPipelineTests.swift]: `closeHeaders` should maybe map to an empty `Data` automatically?

File: Sources/HttpPipeline/Middleware.swift [hunk 4]
```diff
@@ -97,7 +98,7 @@ public func send(_ data: Data?) -> Middleware<BodyOpen, BodyOpen, Data?, Data?>
     }
 
     return .init(
-      data: conn.data,
+      data: concatenatedData,
```
Comment (stephencelis) [inline Sources/HttpPipeline/Middleware.swift]: Failure to set this prevented `send` from working. Not sure if this is the right fix though? Feeling our data model may not be quite what we want it to be here.

**Comment by stephencelis** [comment]
Gonna merge this for now, but definitely reveals that incremental middleware (anything that chunks a payload over time) needs some work.

**Pull request merged**